src/pull.rs

```diff
@@ -83,108 +85,138 @@ impl PullClient {
+        layer: OciDescriptor,
+        diff_id: String,
+        decrypt_config: &Option<&str>,
+        layer_data: Vec<u8>,
+        ms: Arc<Mutex<MetaStore>>,
+    ) -> Result<LayerMeta> {
+        let mut out: Vec<u8> = Vec::new();
+        let plaintext_layer: Vec<u8>;
+
+        let mut layer_meta = LayerMeta::default();
+        let mut media_type_str: &str = layer.media_type.as_str();
+
+        let decryptor = Decryptor::from_media_type(&layer.media_type);
+
+        if decryptor.is_encrypted() {
+            if let Some(dc) = decrypt_config {
+                plaintext_layer = decryptor
+                    .get_plaintext_layer(&layer, layer_data, dc)
+                    .await?;
+                media_type_str = decryptor.media_type.as_str();
+                layer_meta.encrypted = true;
+            } else {
+                return Err(anyhow!("decrypt_config is None"));
+            }
+        } else {
+            plaintext_layer = layer_data;
+        }
+
+        let layer_db = &ms.lock().await.layer_db;
+
+        if let Some(layer_meta) = layer_db.get(&layer.digest) {
+            return Ok(layer_meta.clone());
+        }
+
+        layer_meta.decoder = Compression::try_from(media_type_str)?;
+
+        if layer_meta.decoder == Compression::Uncompressed {
+            layer_meta.uncompressed_digest = layer.digest.clone();
+            layer_meta.compressed_digest = layer.digest.clone();
+        } else {
+            layer_meta.compressed_digest = layer.digest.clone();
+            layer_meta
+                .decoder
+                .decompress(plaintext_layer.as_slice(), &mut out)?;
+
+            if diff_id.starts_with(DIGEST_SHA256) {
+                layer_meta.uncompressed_digest =
+                    format!("{}:{:x}", DIGEST_SHA256, sha2::Sha256::digest(&out));
+            } else if diff_id.starts_with(DIGEST_SHA512) {
+                layer_meta.uncompressed_digest =
+                    format!("{}:{:x}", DIGEST_SHA512, sha2::Sha512::digest(&out));
+            } else {
+                return Err(anyhow!("unsupported digest format: {}", diff_id));
+            }
+        }
+
+        // uncompressed digest should equal to the diff_ids in image_config.
+        if layer_meta.uncompressed_digest != diff_id {
+            return Err(anyhow!(
+                "unequal uncompressed digest {:?} config diff_id {:?}",
+                layer_meta.uncompressed_digest,
+                diff_id
+            ));
+        }
+
+        let store_path = format!(
+            "{}/{}",
+            self.data_dir.display(),
+            &layer.digest.to_string().replace(':', "_")
+        );
+
+        let destination = Path::new(&store_path);
+
+        if let Err(e) = unpack(out, destination) {
+            fs::remove_dir_all(destination)?;
+            return Err(e);
+        }
+
+        layer_meta.store_path = destination.display().to_string();
+
+        Ok(layer_meta)
+    }
+}
+
+impl TryFrom<&str> for Compression {
```

Comment:
Can we move the implementation and tests of `TryFrom` for `Compression` to https://github.com/confidential-containers/image-rs/blob/main/src/decoder/mod.rs ?

Reply:
Good call - done.